File: packages/java/src/main/java/io/opensms/Inbound.java
package io.opensms;

import io.opensms.models.InboundMessage;
import io.opensms.models.Message;
import io.opensms.models.Page;

import java.util.Map;

/** Messages received on your numbers. Accessed as {@code client.inbound()}. */
public final class Inbound {

    private final ApiTransport t;

    Inbound(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of inbound messages. */
    public Page<InboundMessage> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<InboundMessage> list(ListParams params) {
        return t.get("/v1/inbound", params.toQuery(), ApiTransport.pageOf(InboundMessage.class));
    }

    /**
     * Reply to an inbound message (live keys only).
     *
     * @param id   inbound message id.
     * @param text reply text.
     * @return the outbound message.
     */
    public Message reply(String id, String text) {
        return reply(id, text, null);
    }

    /**
     * @param id      inbound message id.
     * @param text    reply text.
     * @param options per-call options.
     * @return the outbound message.
     */
    public Message reply(String id, String text, RequestOptions options) {
        Map<String, Object> body = Wire.map();
        body.put("text", Wire.require(text, "text"));
        return t.json("POST", "/v1/inbound/" + Wire.id(id, "id") + "/reply", body,
                ApiTransport.idempotencyKey(options), ApiTransport.type(Message.class));
    }
}

File: packages/php/src/Resources/Inbound.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `inbound` resource: messages received on your numbers, and replies.
 * Accessed as `$client->inbound`.
 */
final class Inbound
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function list(array $params = []): Page
    {
        return Models::page($this->transport->request('GET', '/v1/inbound', Models::map($params, Models::PAGE_QUERY, [], true)));
    }

    /**
     * Reply to an inbound message. Returns the sent Message (live keys only).
     *
     * @param array{text: string} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function reply(string $id, array $params, array $options = []): array
    {
        $body = Models::map($params, Models::INBOUND_REPLY, ['text']);

        return Models::object($this->transport->request('POST', '/v1/inbound/' . Models::pathParam($id) . '/reply', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }
}
